File: prop_clingo/propagator_clingo_c/propagator_clingo.cpp

#include <clingo.h>
#include <stdlib.h>
#include <stdio.h>
#include <string>
#include <assert.h>
#include "../../utility.h"
#include "../../amosum.h"
#include <sstream>
#include <iostream>
#include <vector>
#include <limits>
#include "propagator_clingo.h"


bool PropagatorClingo::init(clingo_propagate_init_t *_init){

    PropagatorClingoInitializer::get_instance()->init(_init, *this);

    for (size_t i = 0; i < PropagatorClingoInitializer::get_instance()->nt; i++){
        AmoSumPropagator* propagator = new AmoSumPropagator(atomNames, param, propagation_phase, ge, choice_cons, solver = AmoSumPropagator::CLINGO);
        propagator->map_plit_slit = map_plit_slit ;        
        this->propagators.push_back(propagator);
    }
    std::vector<clingo_literal_t> to_watch_plit;

    for (size_t i = 0; i < PropagatorClingoInitializer::get_instance()->nt; i++) to_watch_plit = AmoSumInitializer::get_instance()->getLiterals(*PropagatorClingoInitializer::get_instance()->lits, this->propagators[i]) ;
    
    size_t max_clause_size = this->propagators[0]->N;
    clause_clingo = new clingo_literal_t[max_clause_size];

    for(clingo_literal_t plit: to_watch_plit){
        clingo_literal_t slit = (*map_plit_slit)[plit];
        update_map_value_vector(map_slit_plit_watched, slit, plit);
        handle_error(clingo_propagate_init_add_watch(_init, slit));
    }


    std::vector<clingo_literal_t> S_plit ;
    for (size_t i = 0; i < PropagatorClingoInitializer::get_instance()->nt; i++) S_plit = propagators[i]->simplifyAtLevelZero(true);
    
    if (S_plit.size() == 1 and S_plit[0] == BOTTOM){ 
        bool result ; 
        handle_error(clingo_propagate_init_add_clause((clingo_propagate_init*) _init, NULL, 0, &result));
        return true; 
    }// inconsistent

    add_clauses_propagated_lits(_init, S_plit, 0, true);
    return true ;
}

bool PropagatorClingo::add_clauses_propagated_lits(void *control, const std::vector<clingo_literal_t>& S_plit, int dl, bool init=false){
    
    int td ;
    init ? td = 0 : td = clingo_propagate_control_thread_id((clingo_propagate_control*) control); 

    
    AmoSumPropagator* prop = propagators[td];
    for(int si = 0 ; si < S_plit.size();  ++si){
        clingo_literal_t plit = S_plit[si];
        const std::vector<clingo_literal_t>* R_plit = dl > 0 ? prop->getReasonForLiteral(plit) : nullptr;
        size_t clause_size = (dl > 0 ? R_plit->size() : 0) + 1 ;
        clingo_literal_t* clause = clause_clingo;
        clingo_literal_t slit = (*map_plit_slit)[plit];
        clause[0] = slit ; 
        assert(clause_size <= this->propagators[0]->N);
        for (size_t i = 1; i < clause_size; i++) {
            clingo_literal_t r_plit =  (*R_plit)[i-1];
            clause[i] = (*map_plit_slit)[r_plit];
        }

        bool result_add_clause;
        init ? handle_error(clingo_propagate_init_add_clause((clingo_propagate_init*) control, clause, clause_size, &result_add_clause)) :
        handle_error(clingo_propagate_control_add_clause((clingo_propagate_control*) control, clause, clause_size, clingo_clause_type_learnt, &result_add_clause)) ;

        // propagation must return immediately, there is a conflict
        if (not result_add_clause){
            // debugf("conflict add clause");
            return true ;
        }


        bool result_propagate;
        init ? handle_error(clingo_propagate_init_propagate((clingo_propagate_init*) control, &result_propagate)) :
        handle_error(clingo_propagate_control_propagate((clingo_propagate_control*)control, &result_propagate)) ;
        
        if (!result_propagate){ 
            // propagation must return immediately, a conflict has been raised 
            return true ;
        }   
    }
    return false ;
}   

bool PropagatorClingo::propagate(clingo_propagate_control_t *control, const clingo_literal_t *changes, size_t size){
    const clingo_assignment_t *assignment = clingo_propagate_control_assignment(control);
    int dl = clingo_assignment_decision_level(assignment);
    int td; 
    dl == 0 ? td = 0 : td = clingo_propagate_control_thread_id(control) ; 
    AmoSumPropagator* prop = propagators[td];
    prop->control = control ;
    print_propagate(this, changes, size, control, dl, false, false);


    for (size_t i = 0; i < size; i++)
    {
        clingo_literal_t slit = changes[i];
        std::vector<clingo_literal_t> plit_list = map_slit_plit_watched[slit];
        
        for(clingo_literal_t plit: plit_list){
            const std::vector<clingo_literal_t>* S_plit = prop->onLiteralTrue(plit, dl); // handled internally 
            if (S_plit != nullptr && add_clauses_propagated_lits(control, *S_plit, dl, false)){ 
                // debugf("conflict");
                return true;
            }
                  
        }
    }

    return true;
    
}



std::string PropagatorClingo::compute_changes_str(const clingo_literal_t *changes, size_t size, int td){
    std::vector<std::string> changes_name_vec ; 
    for (size_t i = 0; i < size; i++)
    {
        clingo_literal_t slit = changes[i] ;
        for(clingo_literal_t plit: map_slit_plit_watched[slit]){
            assert(atomNames != NULL);
            std::string name = get_name(atomNames, plit);
            std::string res = "[" + name + ", plit: " + std::to_string(plit) + ", slit: " +  std::to_string(slit) + "]";
            changes_name_vec.push_back(res);
        }
    }

    return vector_to_string(changes_name_vec,"");
}




void PropagatorClingo::undo(clingo_propagate_control_t *control, const clingo_literal_t *changes, size_t size){
   
    const clingo_assignment_t *assignment = clingo_propagate_control_assignment(control);
    int dl = clingo_assignment_decision_level(assignment);
    int td; 
    dl == 0 ? td = 0 : td = clingo_propagate_control_thread_id(control) ; 
    AmoSumPropagator* prop = propagators[td];

    std::vector<clingo_literal_t> plit_list;
   
    
    for (size_t i = 0; i < size; i++)
    {
        clingo_literal_t slit = changes[i];
        extend_vector(plit_list, map_slit_plit_watched[slit]);
    }
    print_undo(this, changes, size, control, dl, td, false, false);
    prop->onLiteralsUndefined(plit_list, false);
}




 

File: amoclingo/propagator_clingo_py/propagator_clingo.py
#!/home/s.fiorentino/miniconda3/bin/python3
import sys
import os
from typing import Sequence
# adding the root path
import clingo
from clingo.symbol import Number, Function
from clingo.control import Control
from clingo.symbol import Function

from amosum_initializer import AmoSumInitializer

sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from amosum import *
import amosum
'''
    Invariants: 
'''
import time 
from amoclingo.propagator_clingo_py.propagator_clingo_initializer import *

class PropagatorClingo(clingo.Propagator):

    def __init__(self, sys_parameters: dict[str: str], propagation_phase: Callable[[Group, AmoSumPropagator], List[int]], ge: bool, choice_cons: str) -> None:
        super().__init__()
        self.sys_parameters = sys_parameters
        self.propagation_phase = propagation_phase
        self.ge = ge
        self.choice_cons = choice_cons
        self.solver = AmoSumPropagator.CLINGO

    def init(self, _init: clingo.PropagateInit) -> None:
        PropagatorClingoInitializer.get_instance().init(_init, self)

        self.propagators = []
        for i in range(PropagatorClingoInitializer.get_instance().nt):
                propagator = amosum.AmoSumPropagator(atomNames=self.atomNames, sys_parameters=self.sys_parameters,
                                      propagation_phase=self.propagation_phase, ge=self.ge, choice_cons=self.choice_cons, solver=AmoSumPropagator.CLINGO)
                propagator.map_plit_slit = self.map_plit_slit
                self.propagators.append(propagator)

        for i in range(PropagatorClingoInitializer.get_instance().nt): to_watch_plit = AmoSumInitializer.get_instance().getLiterals(PropagatorClingoInitializer.get_instance().lits, self.propagators[i])
        
        # arrived here 
        map_slit_plit_watched = {}
        for plit in to_watch_plit:
            slit = self.map_plit_slit[plit]
            map_slit_plit_watched.setdefault(slit,[])
            map_slit_plit_watched[slit].append(plit)
            # debug(f"watching: {slit} {get_name(self.atomNames,plit)}", force_print=True)
            _init.add_watch(literal=slit)
            
            
        self.map_slit_plit_watched = map_slit_plit_watched

        # debug("Propagation ad level 0 started")
        for i in range(PropagatorClingoInitializer.get_instance().nt):
            S_plit = self.propagators[i].simplifyAtLevelZero(delete_lits=True)
            
        # print_derivation(atomNames=self.atomNames, S=S_plit)
        if S_plit == [1] or self.add_clauses_propagated_lits(control=_init, S_plit=S_plit, dl = 0) or not _init.propagate():
            # adding empty clause
            _init.add_clause([])
            return
        
        

    def add_clauses_propagated_lits(self, control: clingo.PropagateControl | clingo.PropagateInit, S_plit, dl):
        td = 0 if isinstance(control, clingo.PropagateInit) else control.thread_id
        prop = self.propagators[td]

        for si in range(len(S_plit)):
            plit = S_plit[si]
            try:
                R_plit = prop.getReasonForLiteral(plit)

                slit  = self.map_plit_slit[plit]
                # first part of the clause is the reason
                clause = [self.map_plit_slit[plit_r] for plit_r in R_plit] 
                
                # the last literal is the implied literal (undefined or false)
                clause.append(slit)
                
                if not control.add_clause(clause):
                    # propagation must return immediately, a conflict has been raised
                    return True
                
                if not control.propagate():
                    return True
            except Exception as e:
                raise e
        return False
    

    def propagate(self, control: clingo.PropagateControl, changes: Sequence[int]) -> None:
        try:
            dl = control.assignment.decision_level
            td = 0 if dl == 0 else control.thread_id
            prop = self.propagators[td]
            prop.control = control
            # print_propagate(self, changes=changes, control=control, dl=dl, force_print=True)

            for ci in range(len(changes)):
                slit = changes[ci]
                plit_list = self.map_slit_plit_watched[slit]
                for plit in plit_list:
                    # propagated plits
                    S_plit = []
                    # propagating program literal
                    S_plit = prop.onLiteralTrue(plit, dl)
                    # adding clauses for propagated literals S_plit
                    if self.add_clauses_propagated_lits(control=control, S_plit=S_plit, dl = dl):
                        return 
            
            
        except Exception as e:
            debug(e, force_print=True)
            tb = e.__traceback__
            while tb:
                print(f"File: {tb.tb_frame.f_code.co_filename}, Line: {tb.tb_lineno}, Function: {tb.tb_frame.f_code.co_name}", file=sys.stderr)
                tb = tb.tb_next
            raise e

    def undo(self, thread_id: int, assignment: clingo.Assignment, changes: Sequence[int]) -> None:
        prop = self.propagators[thread_id]
        plit_list = []
        for slit in changes:
            for plit in self.map_slit_plit_watched[slit]:
                plit_list.append(plit)
        print_undo(self, changes=changes, thread_id=thread_id)
        try:
            prop.onLiteralsUndefined(*plit_list, wasp=False)
        except Exception as e:
            debug(e, force_print=True)
            tb = e.__traceback__
            while tb:
                print(f"File: {tb.tb_frame.f_code.co_filename}, Line: {tb.tb_lineno}, Function: {tb.tb_frame.f_code.co_name}", file=sys.stderr)
                tb = tb.tb_next
            raise e

    def compute_changes_str(self, changes, thread_id):
        changes_str = []
        for slit in changes:
             for plit in self.map_slit_plit_watched[slit]:
                 changes_str.append((get_name(atomNames=self.atomNames, lit = plit), self.map_slit_plit_watched[slit], slit))
        return changes_str
    
    
    def createClingoInterpretation(self, slit_list: List[int], assignment: clingo.Assignment):
        I_clingo = SymmetricFunction(self.propagators[0].N)
        for slit in slit_list:
            plit = self.map_slit_plit_watched[slit]
            I_clingo[plit] = assignment.value(slit)
        return I_clingo